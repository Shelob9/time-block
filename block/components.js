import { DateTimePicker,BaseControl,SelectControl } from '@wordpress/components';
import { __experimentalGetSettings } from '@wordpress/date';
import allTimeZones from "./allTimeZones";
import { useMemo } from "@wordpress/element";
import  {
    findTimeZone, getZonedTime, 
} from 'timezone-support';

/**
 * Choose the primary time/date to convert from.
 * 
 * @see https://developer.wordpress.org/block-editor/components/date-time/ 
 */
export const PrimaryTimePicker = ({
   date,onChange 
}) => {
    const settings = __experimentalGetSettings();

    const is12HourTime = /a(?!\\)/i.test(
        settings.formats.time
            .toLowerCase() // Test only the lower case a
            .replace( /\\\\/g, '' ) // Replace "//" with empty strings
            .split( '' ).reverse().join( '' ) // Reverse the string and test for "a" not followed by a slash
    );
 
    return (
        <BaseControl
            id="primaryTime"
            label="Primary Time"
            help="Enter some text"
        >
            <DateTimePicker
                currentDate={ date }
                onChange={ ( date ) => onChange( date  ) }
                is12Hour={ is12HourTime }
            />
        </BaseControl>
    );
};

/**
 * Choose a timezone 
 */
export const TimeZoneChooser = ({value,onChange}) => {
    return (
        <SelectControl
            label={ 'Select TimeZone' }
            value={ value }
            onChange={ onChange}
            options={ allTimeZones }
        />
    )
}


/**
 * Render a time in another timezone
 */
export const ConvertedTime = ({ primaryTime, timeZone }) => {
    const time = useMemo(() => {
        const zoned = findTimeZone(timeZone);
        const zonedTime = getZonedTime(new Date(primaryTime), zoned);
        const { year, month, day, hours, minutes, seconds } = zonedTime
        return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`
    }, [primaryTime, timeZone]);

    const zone = useMemo(() => `${timeZone.split('/')[1].replace( '_', '')}`,[timeZone]);
    return (
        <div>
            <span>{time}</span> - <span>{zone}</span>
        </div>
    )
}

/**
 * Render one time in multiple timezones
 */
export const ConvertedTimes = ({ primaryTime, timeZones }) => {
    return (
        <div>
            {timeZones.map(timeZone => (<ConvertedTime key={timeZone} primaryTime={primaryTime} timeZone={timeZone}/>))}
        </div>
    )
}