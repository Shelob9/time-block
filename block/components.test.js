import { ConvertedTime,ConvertedTimes } from "./components";
import { render, } from "@testing-library/react";

test.skip('ConvertTime', () => {
    let primaryTime = "2021-03-10T14:30:00";
    let timeZone = "America/New_York"
    let { container } = render(<ConvertedTime {...{ primaryTime, timeZone }} />);
    expect(container).toMatchSnapshot();
});

test.skip('ConvertedTimes', () => {
    let primaryTime = "2021-03-10T14:30:00";
    let timeZones = ["America/New_York", "America/Los_Angeles"];
    let { container } = render(<ConvertedTimes {...{ primaryTime, timeZones }} />);
    expect(container).toMatchSnapshot();
});
