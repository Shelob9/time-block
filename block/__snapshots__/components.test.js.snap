// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConvertTime 1`] = `
<div>
  <div>
    <span>
      2021-3-10 14:30:0
    </span>
     - 
    <span>
      NewYork
    </span>
  </div>
</div>
`;

exports[`ConvertedTimes 1`] = `
<div>
  <div>
    <div>
      <span>
        2021-3-10 14:30:0
      </span>
       - 
      <span>
        NewYork
      </span>
    </div>
    <div>
      <span>
        2021-3-10 11:30:0
      </span>
       - 
      <span>
        LosAngeles
      </span>
    </div>
  </div>
</div>
`;
